libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs

```diff
@@ -487,8 +506,18 @@ public async Task<JToken> ResolveRefAsync(JToken refToken, SourceContext sourceC
         /// <returns>The resulting <see cref="JToken"/> and <see cref="SourceRange"/> for the requested resource.</returns>
         internal async Task<(JToken, SourceRange)> ReadTokenRangeAsync(Resource resource, SourceContext sourceContext, bool advanceJsonReader = false)
         {
-            var text = await resource.ReadTextAsync().ConfigureAwait(false);
-            using (var readerText = new StringReader(text))
+            if (!string.IsNullOrEmpty(resource.FullName))
+            {
+                if (_resourceTokenCache.TryGetValue(resource.FullName, out (JToken, SourceRange) cachedResult))
```

Comment:
Nit: you can actually name the pieces of the pair:

`out (JToken Token, SourceRange SourceRange)` so that then you reference it in the code as `cachedResult.SourceRange`